### PR TITLE
CI update

### PR DESCRIPTION
**Pull request by franziskuskiefer**
Update the CI to 
* add audit
* add clippy
* add rustfmt
* add caching
* have two separate steps for build and tests

### CODE REVIEW

**Review by raphaelrobert — APPROVED** [review]

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -18,14 +18,56 @@ jobs:
...
+      with:
+        path: |
+          ~/.cargo/registry
+          ~/.cargo/git
+          target
+        key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
     - name: Build
       run: cargo build --verbose
+    - name: Build Release
+      run: cargo build --release --all-features --verbose
```
Comment (kkohbrok) [inline .github/workflows/main.yml:34]: Why not build `release` and `debug` in one command?

**Review by kkohbrok — COMMENTED** [review]
Looks good, just one question out of curiosity.

File: .github/workflows/main.yml (hunk 1 above)
Comment (franziskuskiefer) [reply .github/workflows/main.yml:34]: Not sure how you'd do that. Also, the build processes are completely independent.

Comment (kkohbrok) [reply .github/workflows/main.yml:34]: Ah, my bad. I somehow thought `--release` was a target, but it's actually a compilation option.

**Pull request merged**